feat(ext/telemetry): add gRPC protocol support for OTLP exporter

7Hazard opened the pull request:
Add gRPC as a supported protocol for the OpenTelemetry OTLP
exporter, in addition to the existing `http/protobuf` and
`http/json` protocols. This uses a lightweight custom
implementation built on the existing `HyperClient` with gRPC
wire framing (5-byte length-prefixed protobuf over HTTP/2),
avoiding heavy dependencies like `tonic`.

```sh
OTEL_EXPORTER_OTLP_PROTOCOL=grpc \
  deno run --unstable-otel main.ts
```

Features:
- `GrpcSpanExporter`, `GrpcLogExporter`, `GrpcMetricExporter`
- Proper `grpc-status` handling (both headers and trailers)
- `OTEL_EXPORTER_OTLP_COMPRESSION=gzip` support
- Per-signal endpoint/headers env vars
  (`OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`, etc.)
- Enforced HTTP/2 via `HyperClient::new_h2()`
- Response body capped at 1 MB
- gRPC export errors surfaced via `log::error!`

Closes #27920

7Hazard committed:
feat(ext/telemetry): add gRPC support for OTLP exporter

This adds gRPC protocol support to the OpenTelemetry exporter,
allowing users to set OTEL_EXPORTER_OTLP_PROTOCOL=grpc.

Closes #27920

7Hazard committed:
feat(ext/telemetry): Add test files for otlp grpc

7Hazard committed:
feat(ext/telemetry): Set up otel grpc server for tests

bartlomieju commented:
Thanks for the PR @7Hazard, let me know if you need any help getting it landed

7Hazard commented:
@bartlomieju Unfortunately got stuck on trying to get the otel exported through grpc successfully in the test suites.
I'm afraid it might have to do something with incompatibility of how the tokio runtime is used in Deno and Tonic (what the Rust otel crate uses to drive grpc comms), based on what I've read in various GitHub issues.
https://github.com/hyperium/tonic/issues/942#issuecomment-1313396286
I'm not fully versed in how Deno handles the Tokio runtimes, so some info about that or leads on how to proceed would be great.
Kind of limited on time but I'll try to get this sorted when I find the time.

bartlomieju commented:
@7Hazard sorry for a slow response, regarding:

> I'll see if I could figure it out, but makes me think there might be some difficulty with starting a server using npm:@grpc/grpc-js or I've not implemented the otel proto loading properly.

Yeah, Deno doesn't correctly handle `npm:@grpc/grpc-js` server side - see https://github.com/denoland/deno/issues/23714

lucasff commented:
This should be prioritized, please.
we're running Deno at scale and GRPC is arguably better in terms of performance and cost.

bartlomieju commented:
@lucasff can you shoot me an email to bartek at deno.com so I can learn a bit more about your needs?

bartlomieju committed:
merge main and reimplement gRPC OTLP exporter without tonic

Merges latest main (1401 commits) and replaces the original tonic-based
gRPC exporter with a lightweight custom implementation that uses the
existing HyperClient with gRPC wire framing (5-byte prefix over h2).

- Remove grpc-tonic feature from opentelemetry-otlp dependency
- Add opentelemetry-proto + prost as direct dependencies for proto types
- New grpc_exporter.rs: GrpcSpanExporter, GrpcLogExporter, GrpcMetricExporter
- Custom protobuf decoder (proto_decode.ts) for test server, no npm deps
- gRPC test variants (basic_grpc, metric_grpc, events_grpc)

bartlomieju committed:
Merge remote-tracking branch 'origin/main' into otlp-grpc

# Conflicts:
#	.github/workflows/node_compat_test.generated.yml
#	.github/workflows/node_compat_test.ts
#	Cargo.lock
#	Cargo.toml
#	cli/args/mod.rs
#	cli/tools/x.rs
#	ext/cache/lib.rs
#	ext/node/lib.rs
#	ext/node/ops/stream_wrap.rs
#	ext/node/ops/tcp_wrap.rs
#	ext/node/ops/tls_wrap.rs
#	ext/node/ops/vm.rs
#	ext/node/polyfills/_fs/_fs_lstat.ts
#	ext/node/polyfills/_http_client.js
#	ext/node/polyfills/_http_outgoing.ts
#	ext/node/polyfills/_process/process.ts
#	ext/node/polyfills/_tls_wrap.js
#	ext/node/polyfills/async_hooks.ts
#	ext/node/polyfills/fs.ts
#	ext/node/polyfills/http2.ts
#	ext/node/polyfills/https.ts
#	ext/node/polyfills/internal/buffer.mjs
#	ext/node/polyfills/internal/crypto/_randomBytes.ts
#	ext/node/polyfills/internal/crypto/cipher.ts
#	ext/node/polyfills/internal/crypto/diffiehellman.ts
#	ext/node/polyfills/internal/crypto/hash.ts
#	ext/node/polyfills/internal/errors.ts
#	ext/node/polyfills/internal/timers.mjs
#	ext/node/polyfills/internal_binding/tls_wrap.ts
#	ext/node/polyfills/net.ts
#	ext/node/polyfills/process.ts
#	ext/node/polyfills/timers.ts
#	ext/node_crypto/lib.rs
#	ext/node_sqlite/database.rs
#	ext/telemetry/lib.rs
#	libs/core/uv_compat/pipe.rs
#	libs/core/uv_compat/tcp.rs
#	runtime/permissions/lib.rs
#	tests/node_compat/config.jsonc
#	tests/node_compat/mod.rs
#	tests/node_compat/slack.ts
#	tests/specs/cli/otel_basic/__test__.jsonc
#	tests/unit_node/buffer_test.ts
#	tests/unit_node/crypto/crypto_hash_test.ts
#	tests/unit_node/fs_test.ts
#	tests/unit_node/http2_test.ts
#	tests/unit_node/http_test.ts
#	tests/unit_node/process_test.ts
#	tests/unit_node/tls_test.ts
#	tests/wpt/runner/expectations/service-workers.json
#	tools/lint_plugins/no_deno_api_in_polyfills.ts

bartlomieju committed:
fix(ext/telemetry): fix protobuf decoder for gRPC OTLP tests

Fix several issues in the proto_decode.ts protobuf decoder used by the
gRPC OTLP test suite:

- Decode `double` fields as IEEE 754 floats instead of integer bit
  patterns (was showing raw u64 representations like "4572414629676717179"
  instead of 0.005)
- Decode `float` fields using DataView.getFloat32 for correct IEEE 754
  interpretation
- Serialize `int64`/`sint64` values as JSON strings to match the Rust
  serde AnyValue custom serializer (e.g., `"intValue": "200"`)
- Treat proto3 synthetic oneofs (prefixed with `_`) as regular fields,
  fixing field ordering for `sum`, `min`, `max` in HistogramDataPoint
- Add default `"0"` string for unset int64/sint64 fields

bartlomieju committed:
Merge remote-tracking branch 'origin/main' into otlp-grpc

bartlomieju committed:
Merge remote-tracking branch 'origin/main' into otlp-grpc

bartlomieju committed:
fix: address gRPC OTLP exporter review feedback

Blockers:
- Fix grpc-status trailer handling: add grpc_request() method on
  HyperClient that preserves HTTP/2 trailers via Collected::trailers()
  instead of discarding them with to_bytes(). grpc_send now checks
  grpc-status in both initial headers (Trailers-Only form) and
  trailers (normal form).
- Enforce HTTP/2 for gRPC: add HyperClient::new_h2() that sets
  http2_only(true) on the hyper client builder. gRPC transport now
  uses this, ensuring HTTP/2 is used even for plaintext endpoints.

Non-blocking:
- Fix per-signal endpoint env var precedence: per-signal vars
  (OTEL_EXPORTER_OTLP_TRACES_ENDPOINT etc.) now take precedence
  over the generic OTEL_EXPORTER_OTLP_ENDPOINT, per the OTLP spec.
  Each signal (traces/metrics/logs) checks its own endpoint var.
- Add per-signal headers env var support: OTEL_EXPORTER_OTLP_TRACES_HEADERS
  etc. are merged with (and override) the generic headers.
- Remove single-element array in grpc_headers (was a leftover).
- Use .unwrap() instead of .expect() for prost encode (unreachable).

bartlomieju committed:
chore: rustfmt

bartlomieju committed:
Merge branch 'main' into otlp-grpc

bartlomieju committed:
test(ext/telemetry): add regression test for gRPC trailer path

Add a test server (grpc_trailer_main.ts) using node:http2 that sends
grpc-status in HTTP/2 trailers instead of initial response headers.
This exercises the trailer-reading code path in grpc_send, which was
previously only tested via the Trailers-Only form.

bartlomieju committed:
Merge remote-tracking branch '7hazard/otlp-grpc' into otlp-grpc

bartlomieju committed:
fix(ext/telemetry): use clearEnv in otel test child processes

Prevent host environment variables (e.g.
OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE) from leaking into
the test child process and causing spurious failures.

bartlomieju committed:
test(ext/telemetry): add gRPC trailers error test case

Add grpc_trailers_error test that sends grpc-status: 14 (UNAVAILABLE)
in HTTP/2 trailers. The server skips data collection on error, so the
test verifies the output has empty data arrays, confirming the error
path was exercised and the process exits cleanly.

bartlomieju committed:
fix(ext/telemetry): fix clearEnv breaking otel tests on Windows/CI

Revert clearEnv on main.ts since it strips env vars needed by other
tests (metric_temporality_* need OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE)
and breaks DNS on Windows (missing SystemRoot).

Keep clearEnv + OTEL env filtering on grpc_trailer_main.ts since it
has no existing tests depending on inherited OTEL vars.

bartlomieju committed:
fix(ext/telemetry): surface gRPC export errors on stderr

Log gRPC export errors via eprintln so they are visible to users and
testable. Update grpc_trailers_error test to assert the error message
appears, making it a true regression test: if check_grpc_status is
removed, the error message disappears and the test fails.

Also always collect data on the server side (regardless of error
status) and extract env filter into a helper per review feedback.

bartlomieju committed:
fix: use log::error instead of eprintln, collapse nested if

Replace eprintln! with log::error! to satisfy clippy::print_stderr,
and collapse nested if per clippy::collapsible_if. Set DENO_LOG=error
on the child process in error tests so the log message is visible.

bartlomieju commented:
All items from the review are now addressed.

Already fixed in prior commits on this branch:
- gRPC trailers handling (grpc-status in both headers and trailers)
- HTTP/2 enforcement via `new_h2()`
- Per-signal endpoint/headers env var precedence

Newly addressed in 7c7bb2fa8b:
1. **`load_private_keys().next().unwrap()` panic** → `.ok_or_else()` with descriptive error
2. **`parse_otlp_timeout` accepts `0`** → `if millis > 0` guard, falls back to 10s default
3. **Endpoint scheme handling** — bare `host:port` now gets scheme from `OTEL_EXPORTER_OTLP_INSECURE`
4. **`OTEL_EXPORTER_OTLP_COMPRESSION`** — `gzip` honored via flate2 + `grpc-encoding` header; unsupported values warn once
5. **`user-agent` + `grpc-timeout` headers** — `OTel-OTLP-Exporter-Deno/<ver>` and `<ms>m`
6. **Invalid header warnings** — `log::warn!` on bad header names/values
7. **Response body size cap** — 1 MB limit via `http_body_util::Limited` (both HTTP and gRPC paths)

bartlomieju committed:
fix: address remaining gRPC exporter review feedback

- Replace load_private_keys().next().unwrap() with proper error
- Reject OTEL_EXPORTER_OTLP_TIMEOUT=0, fall back to 10s default
- Handle bare host:port endpoints via OTEL_EXPORTER_OTLP_INSECURE
- Honor OTEL_EXPORTER_OTLP_COMPRESSION=gzip with flate2
- Add user-agent and grpc-timeout headers to gRPC requests
- Warn on invalid OTEL_EXPORTER_OTLP_HEADERS entries
- Cap response body collection at 1 MB (both HTTP and gRPC paths)
